matrix/src/armory/matrix/matrix.py

```diff
@@ -0,0 +1,267 @@
+        allowable values for each argument. The yielded value of the generator
+        is a key-value mapping of arguments corresponding to a row of the
+        matrix.
+
+        If an argument to the generator is a callable, it will be called with
+        the preceding keyword parameter values for the current row. The return
+        of the callable is the value--or iterable--for the keyword, from which
+        to generate additional rows in the matrix.
+    """
+
+    if worker_num is None or num_workers is None:
+        worker_num = 0
+        num_workers = 1
+
+    def _generate(**kwargs) -> Iterable[Mapping[str, Any]]:
+        count = 0
+        for params in product({}, list(kwargs.items())):
+            # Skip over entries that are pruned
+            if prune is not None and prune(**params):
+                continue
+            # Skip over entries when partioning
+            if (count % num_workers) == worker_num:
+                yield params
+            count += 1
+
+    return _generate
+
+
+class Matrix:
+    """
+    A function to be invoked multiple times with parameters from rows of a
+    cartesian product matrix of all possible parameters.
+
+    This class is not intended to be used directly, but is used internally by
+    the `matrix` decorator.
+    """
+
+    def __init__(self, func: Callable, **kwargs):
+        self.func = func
+        self.kwargs = kwargs
+        self.worker_num: Optional[int] = None
+        self.num_workers: Optional[int] = None
+        self.pruner: Optional[Callable[..., bool]] = None
+
+    @property
+    def matrix(self):
+        """The generated matrix of arguments"""
+        return create_matrix(
+            worker_num=self.worker_num,
+            num_workers=self.num_workers,
+            prune=self.pruner,
+        )(**self.kwargs)
+
+    def __call__(self, *args, **kwargs):
+        """
+        Invokes the function once for each row of the matrix. Any given keyword
+        arguments are merged with the keyword arguments from the matrix row.
+
+        Args:
+            *args: Positional arguments to be included with each function
+                invocation
+            **kwargs: Keyword arguments to be merged with the matrix row
+                arguments for each function invocation
+
+        Returns:
+            List of return values from each function invocation
+        """
+        results = []
+        for it in self.matrix:
+            it_args = deepcopy(args)
+            it_kwargs = deepcopy(kwargs)
+            it_kwargs.update(it)
+            try:
+                results.append(self.func(*it_args, **it_kwargs))
+            except Exception as err:
+                results.append(err)
+        return results
+
+    def override(self, **kwargs):
+        """Overrides the input parameter constraints for the matrix."""
+        self.kwargs.update(kwargs)
+        return self
+
+    def partition(self, worker_num: Optional[int], num_workers: Optional[int]):
```

Comment:
As mentioned above, slice notation may be a cleaner way to specify a partition of the matrix rows.